google/data_source_google_service_account.go

```diff
@@ -0,0 +1,84 @@
+package google
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourceGoogleServiceAccount() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceGoogleServiceAccountRead,
+
+		Schema: map[string]*schema.Schema{
+			// Required
+			"account_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			// Optional
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			// Computed
+			"email": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"unique_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"display_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceGoogleServiceAccountRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	// Get the account id as the fully qualified name
+	accountID := d.Get("account_id").(string)
+
+	// If the account id isn't already the fully qualified name
+	if !strings.HasPrefix(accountID, "projects/") {
+
+		// If the account id is an email
+		if strings.Contains(accountID, "@") {
+			accountID = "projects/-/serviceAccounts/" + accountID
+		} else {
+			// Get the project from the resource or fallback to the project
+			// in the provider configuration
+			project, err := getProject(d, config)
+			if err != nil {
+				return err
+			}
+			// If the account id doesn't contain the email, build it
+			accountID = fmt.Sprintf("projects/-/serviceAccounts/%s@%s.iam.gserviceaccount.com", accountID, project)
+		}
+	}
+
+	sa, err := config.clientIAM.Projects.ServiceAccounts.Get(accountID).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Service Account %q", accountID))
+	}
+
+	d.SetId(sa.Name)
+	d.Set("email", sa.Email)
+	d.Set("unique_id", sa.UniqueId)
+	d.Set("project", sa.ProjectId)
+	d.Set("account_id", strings.Split(sa.Email, "@")[0])
```

Comment:
_Probably_ not ideal to set `account_id` - users will be a little surprised if `account_id` changes from the value they set it to, on the off chance they choose to interpolate it.  In a resource this would be a permanent diff (unless we had a `DiffSuppressFunc`), which we try to avoid.  I'm 50/50 on this - what do you think?

Reply:
Hm.  That must have a `DiffSuppressFunc`, then?  I suppose it's better to maintain consistency.

Reply:
OK, I was misusing the service account id for this, the service account resource only accepts a service account id, so I have changed the datasource to mimic this; as such the account_id should always be the same. The resource for service account still has this line, so I left it in, but if it is confusing I can remove it?